staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go

```diff
@@ -0,0 +1,1221 @@
+			},
+		},
+		{name: "container type composition",
+			obj: map[string]interface{}{
+				"obj": map[string]interface{}{
+					"field": "a",
+				},
+				"mapOfMap": map[string]interface{}{
+					"x": map[string]interface{}{
+						"y": "b",
+					},
+				},
+				"mapOfObj": map[string]interface{}{
+					"k": map[string]interface{}{
+						"field2": "c",
+					},
+				},
+				"mapOfListMap": map[string]interface{}{
+					"o": []interface{}{
+						map[string]interface{}{
+							"k": "1",
+							"v": "d",
+						},
+					},
+				},
+				"mapOfList": map[string]interface{}{
+					"l": []interface{}{"e"},
+				},
+				"listMapOfObj": []interface{}{
+					map[string]interface{}{
+						"k2": "2",
+						"v2": "f",
+					},
+				},
+				"listOfMap": []interface{}{
+					map[string]interface{}{
+						"z": "g",
+					},
+				},
+				"listOfObj": []interface{}{
+					map[string]interface{}{
+						"field3": "h",
+					},
+				},
+				"listOfListMap": []interface{}{
+					[]interface{}{
+						map[string]interface{}{
+							"k3": "3",
+							"v3": "i",
+						},
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"obj": objectType(map[string]schema.Structural{
+					"field": stringType,
+				}),
+				"mapOfMap": mapType(mapTypePtr(&stringType)),
+				"mapOfObj": mapType(objectTypePtr(map[string]schema.Structural{
+					"field2": stringType,
+				})),
+				"mapOfListMap": mapType(listMapTypePtr([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k": stringType,
+					"v": stringType,
+				}))),
+				"mapOfList": mapType(listTypePtr(&stringType)),
+				"listMapOfObj": listMapType([]string{"k2"}, objectTypePtr(map[string]schema.Structural{
+					"k2": stringType,
+					"v2": stringType,
+				})),
+				"listOfMap": listType(mapTypePtr(&stringType)),
+				"listOfObj": listType(objectTypePtr(map[string]schema.Structural{
+					"field3": stringType,
+				})),
+				"listOfListMap": listType(listMapTypePtr([]string{"k3"}, objectTypePtr(map[string]schema.Structural{
+					"k3": stringType,
+					"v3": stringType,
+				}))),
+			}),
+			expectCost: map[string]uint64{
+				"self.obj.field == 'a'":                                       4,
+				"self.mapOfMap['x']['y'] == 'b'":                              5,
+				"self.mapOfObj['k'].field2 == 'c'":                            5,
+				"self.mapOfListMap['o'].exists(e, e.k == '1' && e.v == 'd')":  14,
+				"self.mapOfList['l'][0] == 'e'":                               5,
+				"self.listMapOfObj.exists(e, e.k2 == '2' && e.v2 == 'f')":     13,
+				"self.listOfMap[0]['z'] == 'g'":                               5,
+				"self.listOfObj[0].field3 == 'h'":                             5,
+				"self.listOfListMap[0].exists(e, e.k3 == '3' && e.v3 == 'i')": 14,
+			},
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			for validRule, expectedCost := range tt.expectCost {
+				t.Run(validRule, func(t *testing.T) {
```

Comment:
cel compilation is surprisingly slow when running with race detection enabled in tests... making the tests run in parallel helps a lot (see https://github.com/kubernetes/kubernetes/pull/108419/files#diff-a04b2bed11e294dc38c72b900c40d932bd1780663f0301022f3ac61a728464de as an example)

I'd suggest adding this:
```suggestion
	for _, tt := range cases {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			for validRule, expectedCost := range tt.expectCost {
				validRule := validRule
				expectedCost := expectedCost
				t.Run(validRule, func(t *testing.T) {
					t.Parallel()
```